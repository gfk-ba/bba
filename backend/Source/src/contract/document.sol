pragma solidity ^0.4.21;

contract Document {

    string    public data;
    address[] public links;

    constructor(string newData, address[] newLinks) public {
        data = newData;
        links = newLinks;
    }
}

// "", [0xE0f5206BBD039e7b0592d8918820024e2a7437b9]