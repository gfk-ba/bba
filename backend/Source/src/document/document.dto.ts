/**
 * The DTO (data transfer object) used to transport document data from the frontend to the backend.
 */
interface DocumentDto {
  /**
   * The id of the project to which the document belongs.
   * To be discussed: Why is this a string?
   */
  readonly projectId: string;

  /**
   * The kind of document (just a string)
   */
  readonly kind: string;

  /**
   * The hash of the document to store in the blockchain
   */
  readonly hash: string;

  /**
   * The description of the document (given by the user)
   */
  readonly description: string;

  /**
   * The list of linked documents (as a list of ids)
   */
  readonly linkIds: any;

  /**
   * The payment data (e.g. from PayPal)
   */
  readonly payment: any;
}
