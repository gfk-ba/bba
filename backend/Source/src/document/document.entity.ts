/**
 * This files contains the entity class for the document
 */

/**
 * import all decorator from typeorm
 */
import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';

/**
 * The class Document is used with the ORM TypeORM to access data in the database table documents.
 */
@Entity('documents')
export class Document {
  /**
   * The ID (primary key) of the table (autogenerated)
   */
  @PrimaryGeneratedColumn()
  id: number;

  /**
   * The hashed data of the document
   */
  @Column({ type: 'varchar', length: 128 })
  hash: string;

  /**
   * The id of the project to which the document belongs (foreign key)
   */
  @Column({ type: 'int', name: 'project_id' })
  projectId: number;

  /**
   * The kind (type) of the document, must be choosen by the user in the frontend.
   */
  @Column({ type: 'varchar', length: 10 })
  kind: string;

  /**
   * The description of the document, can be provided by the user in the frontend.
   */
  @Column({ type: 'varchar', length: 255 })
  description: string;

  /**
   * The etherum address of the document
   */
  @Column({ type: 'varchar', length: 42, nullable: true })
  address: string;

  /**
   * The current state of the document (e.g. whether the document is saved the blockchain).
   * Could be OK, ERROR, PROGRESS...
   */
  @Column({ type: 'varchar', length: 10, default: 'PROGRESS' })
  state: string;

  /**
   * Details about the state of the document, especially error messages if something went wrong.
   */
  @Column({ type: 'varchar', length: 255, name: 'state_details', default: '' })
  stateDetails: string;

  /**
   * The timestamp when the document was stored in the database. \
   * Note: "timestamptz" is NOT a typo, but prevents SQL errors because timestamp is a reserved word.
   */
  @Column({ type: 'timestamptz', name: 'created_at', nullable: true })
  createdAt: Date;
}
