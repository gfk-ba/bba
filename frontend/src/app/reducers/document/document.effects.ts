import { Inject, Injectable } from '@angular/core';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { timer } from 'rxjs';
import { flatMap, mapTo, takeUntil } from 'rxjs/operators';
import { Environment } from '../../../environments/environment';
import {
  DocumentsListActionTypes,
  FetchDocumentsList,
} from './documents-list/documents-list.actions';

@Injectable()
export class DocumentEffects {
  @Effect()
  pollDocuments = this.actions$.pipe(
    ofType(DocumentsListActionTypes.StartPolling),
    flatMap(() =>
      timer(0, this.env.documentPollingTimer).pipe(
        takeUntil(
          this.actions$.pipe(
            ofType(DocumentsListActionTypes.StopPolling, DocumentsListActionTypes.StartPolling)
          )
        ),
        mapTo(new FetchDocumentsList())
      )
    )
  );

  constructor(private actions$: Actions, @Inject(APP_ENVIRONMENT) private env: Environment) {}
}
