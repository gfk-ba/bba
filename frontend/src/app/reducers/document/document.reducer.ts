import { ActionReducer, combineReducers } from '@ngrx/store';
import { ICurrentDocumentState } from './current-document/current-document.manager';
import { currentDocument } from './current-document/current-document.reducer';
import { DocumentsListActions } from './documents-list/documents-list.actions';
import { IDocumentsListState } from './documents-list/documents-list.manager';
import { documentsList } from './documents-list/documents-list.reducer';

export interface State {
  currentDocument: ICurrentDocumentState;
  documentsList: IDocumentsListState;
}

export const reducer: ActionReducer<State, DocumentsListActions> = combineReducers({
  currentDocument,
  documentsList,
});
