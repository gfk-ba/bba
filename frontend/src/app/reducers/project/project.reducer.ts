import { ActionReducer, combineReducers } from '@ngrx/store';

import { ICurrentProjectState } from './current-project/current-project.manager';
import { currentProject } from './current-project/current-project.reducer';
import { IProjectsListState } from './projects-list/projects-list.manager';
import { projectsList } from './projects-list/projects-list.reducer';

export interface State {
  currentProject: ICurrentProjectState;
  projectsList: IProjectsListState;
}

export const reducer: ActionReducer<State> = combineReducers({
  currentProject,
  projectsList,
});
