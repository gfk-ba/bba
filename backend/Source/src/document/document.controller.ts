/**
 * This files contains the DocumentController
 */

/**
 * import all the stuff we need from NestJS
 */
import {
  Controller,
  Get,
  Post,
  Body,
  UseGuards,
  Req,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
import { DocumentService } from './document.service';
import { Document } from './document.entity';
import { AuthGuard } from '@nestjs/passport';
import { UserDataService } from '../shared/auth/user-data.service';
import * as logger from '../shared/logger';

/**
 * The DocumentController provides methods to find all documents of an user [[findAll]] and to create/store a new document [[createDocument]].
 * All methods are accessible at http://SERVER:PORT/documents
 */
@Controller('documents')
export class DocumentController {
  /**
   * Use dependency injection to get access to the [[DocumentService]] and the [[UserDataService]].
   */
  constructor(
    private readonly documentService: DocumentService,
    private userDataService: UserDataService,
  ) {}

  /**
   * This async function delivers the list of all documents of the user, who makes the call.
   * If user data is missing or the user is unknown, the result will the the http status code 417.
   * This function can be executed by a GET call to http://SERVER:PORT/documents
   * ```
   *   curl -X GET \
   *     http://SERVER:PORT/documents \
   *     -H 'Authorization: Bearer XXXXX' \
   *     -H 'Cache-Control: no-cache'
   * ```
   * This route is restricted by [[AuthGuard]]
   * @param requestData The request data, we need this to extract the user data
   * @returns The list of document of the current user (including links to other documents).
   */
  @Get()
  @UseGuards(AuthGuard())
  async findAll(@Req() requestData): Promise<Document[]> {
    const userId = this.userDataService.getUserData(requestData);
    if (!userId) {
      throw new HttpException(
        'userId missing in DocumentController.findAll',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    const result = await this.documentService.findAll(userId);
    return result;
  }
  /**
   * This async function creates+stores a new document using the request data. The document will be associated to the user, who makes the call.
   * If user data is missing or the user is unknown, the result will the the http status code 417.
   * This function can be executed by a POST call to http://SERVER:PORT/documents
   * ```
   * curl -X POST \
   *    http://SERVER:PORT/documents \
   *   -H 'Authorization: Bearer XXXXX' \
   *   -H 'Cache-Control: no-cache' \
   *   -H 'Content-Type: application/x-www-form-urlencoded' \
   *   -d 'projectId=3&kind=result&hash=12121212&description=testdaten&linkIds=%5B2%2C%204%5D'
   * ```
   * The result should be something like this:
   * ```
   * {
   *        "projectId": 3,
   *        "kind": "result",
   *        "hash": "12121212",
   *        "description": "testdaten",
   *        "address": null,
   *        "id": 4,
   *        "state": "PROGRESS",
   *        "stateDetails": "",
   *        "createdAt": "2018-09-16"
   * }
   * ```
   * This route is restricted by [[AuthGuard]]
   * @param documentDto The data of the new document
   * @param requestData The request data, we need this to extract the user data
   * @returns the newly created document - including the database id.
   */
  @Post()
  @UseGuards(AuthGuard())
  async createDocument(
    @Body() documentDto: DocumentDto,
    @Req() requestData,
  ): Promise<any> {
    logger.log('DocumentController.createDocument - documentDto', documentDto);

    // documentDto should be something like this ....
    // {projectId: "3",
    // kind: "result",
    // hash: "212121",
    // description: "ttttt",
    // linkIds: [2, 4]}
    // payment: { }

    if (!this.documentService.hasValidPaymentData) {
      throw new HttpException(
        'payment receipt is missing',
        HttpStatus.EXPECTATION_FAILED,
      );
    }

    const userId = this.userDataService.getUserData(requestData);

    const savedDocument = await this.documentService.createDocument(
      userId,
      documentDto,
    );

    logger.log(
      'DocumentController.createDocument - savedDocument',
      savedDocument,
    );

    return savedDocument;
  }
}
