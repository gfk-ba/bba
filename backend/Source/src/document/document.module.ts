/**
 * This file contains the [[DocumentModule]] which provides the [[DocumentService]] and [[DocumentController]] to other modules
 */

/**
 * Importing stuff from NestJS and our own code.
 */
import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DocumentService } from './document.service';
import { DocumentController } from './document.controller';
import { Document } from './document.entity';
import { PaymentModule } from '../payment/payment.module';
import { ContractModule } from '../contract/contract.module';
import { AuthModule } from '../shared/auth/auth.module';
import { ProjectModule } from '../project/project.module';
import { LinkModule } from '../link/link.module';

/**
 * This module provides the [[DocumentService]] and [[DocumentController]] to other modules
 */
@Module({
  imports: [
    TypeOrmModule.forFeature([Document]),
    AuthModule,
    ProjectModule,
    LinkModule,
    AuthModule,
    PaymentModule,
    ContractModule,
  ],
  providers: [DocumentService],
  controllers: [DocumentController],
  exports: [DocumentService],
})
export class DocumentModule {}
