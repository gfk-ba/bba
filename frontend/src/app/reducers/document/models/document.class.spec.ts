import { Document } from './document.class';
import { Project } from '../../project/models/project.class';

describe('Document', () => {
  let subject: Document;

  describe('with empty props', () => {
    beforeEach(() => {
      subject = new Document();
    });

    it('creates', () => {
      expect(subject).toBeTruthy();
      expect(subject).toBeInstanceOf(Document);
    });

    it('uses default values', () => {
      expect(subject).toMatchSnapshot();
    });
  });

  describe('with props', () => {
    beforeEach(() => {
      subject = new Document({
        id: '42',
        description: 'The description',
        kind: 'model',
        state: 'ok',
        linkIds: [12],
        projectId: 5,
        project: { id: '5', name: 'Project 1', description: 'fffff', userId: 12 },
        links: [
          {
            id: '12',
            projectId: 5,
            project: { id: '5', name: 'Project 1', description: 'fffff', userId: 12 },
            description: 'Linked description',
            kind: 'custom',
            hash: '12345',
            linkIds: [],
            links: [],
            state: 'ok',
            stateDetails: '',
            address: '',
            createdAt: '',
          },
        ],
      });
    });

    it('creates', () => {
      expect(subject).toBeTruthy();
      expect(subject).toBeInstanceOf(Document);
    });

    it('merges default and given values', () => {
      expect(subject).toMatchSnapshot();
    });

    it('renders the label', () => {
      expect(subject.label).toEqual('#42(Project 1 - Model)');
    });

    it('creates Document instances in links', () => {
      expect(subject.links[0]).toBeInstanceOf(Document);
      expect(subject.links).toMatchSnapshot();
    });

    it('creates Project instance in project', () => {
      expect(subject.project).toBeInstanceOf(Project);
      expect(subject.project).toMatchSnapshot();
    });

    it('calculates graph nodes', () => {
      expect(subject.getGraphNodes()).toMatchSnapshot();
    });

    it('calculates graph edges', () => {
      expect(subject.getGraphEdges()).toMatchSnapshot();
    });

    it('generates kindLabel', () => {
      expect(subject.kindLabel).toEqual('Model');
    });

    it('generates stateLabel', () => {
      expect(subject.stateLabel).toEqual('OK');
    });

    it('generates linksText', () => {
      expect(subject.linksText).toEqual('#12(Project 1 - Custom)');
    });
  });
});
