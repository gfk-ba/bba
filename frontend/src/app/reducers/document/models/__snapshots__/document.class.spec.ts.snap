// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document with empty props uses default values 1`] = `"{\\"id\\":\\"\\",\\"projectId\\":0,\\"description\\":\\"\\",\\"kind\\":\\"data\\",\\"hash\\":\\"\\",\\"linkIds\\":[],\\"links\\":[],\\"project\\":null,\\"state\\":\\"in_progress\\",\\"stateDetails\\":\\"\\",\\"address\\":\\"\\",\\"createdAt\\":\\"\\"}"`;

exports[`Document with props calculates graph edges 1`] = `
Array [
  Object {
    "from": "42",
    "to": "12",
  },
]
`;

exports[`Document with props calculates graph nodes 1`] = `
Array [
  Object {
    "id": "42",
    "label": "#42(Project 1 - Model)",
  },
  Object {
    "id": "12",
    "label": "#12(Project 1 - Custom)",
  },
]
`;

exports[`Document with props creates Document instances in links 1`] = `
Array [
  "{\\"id\\":\\"12\\",\\"projectId\\":5,\\"description\\":\\"Linked description\\",\\"kind\\":\\"custom\\",\\"hash\\":\\"12345\\",\\"linkIds\\":[],\\"links\\":[],\\"project\\":{\\"id\\":\\"5\\",\\"name\\":\\"Project 1\\",\\"description\\":\\"fffff\\",\\"userId\\":12},\\"state\\":\\"ok\\",\\"stateDetails\\":\\"\\",\\"address\\":\\"\\",\\"createdAt\\":\\"\\"}",
]
`;

exports[`Document with props creates Project instance in project 1`] = `"{\\"id\\":\\"5\\",\\"userId\\":12,\\"name\\":\\"Project 1\\",\\"description\\":\\"fffff\\"}"`;

exports[`Document with props merges default and given values 1`] = `"{\\"id\\":\\"42\\",\\"projectId\\":5,\\"description\\":\\"The description\\",\\"kind\\":\\"model\\",\\"hash\\":\\"\\",\\"linkIds\\":[12],\\"links\\":[{\\"id\\":\\"12\\",\\"projectId\\":5,\\"project\\":{\\"id\\":\\"5\\",\\"name\\":\\"Project 1\\",\\"description\\":\\"fffff\\",\\"userId\\":12},\\"description\\":\\"Linked description\\",\\"kind\\":\\"custom\\",\\"hash\\":\\"12345\\",\\"linkIds\\":[],\\"links\\":[],\\"state\\":\\"ok\\",\\"stateDetails\\":\\"\\",\\"address\\":\\"\\",\\"createdAt\\":\\"\\"}],\\"project\\":{\\"id\\":\\"5\\",\\"name\\":\\"Project 1\\",\\"description\\":\\"fffff\\",\\"userId\\":12},\\"state\\":\\"ok\\",\\"stateDetails\\":\\"\\",\\"address\\":\\"\\",\\"createdAt\\":\\"\\"}"`;
